src/Rules.hs

```diff
@@ -44,7 +45,14 @@ buildLib stage pkg = do
             ways <- interpretInContext context getLibraryWays
             libs <- mapM (pkgLibraryFile . Context stage pkg) ways
             docs <- interpretInContext context $ buildHaddock flavour
-            need $ libs ++ [ pkgHaddockFile context | docs && stage == Stage1 ]
+
+            version <- interpretInContext context (getPkgData Version)
+            let topObj = buildPath context -/- "HS" ++
+                         pkgNameString pkg ++ "-" ++ version ++
+                         (waySuffix (way context) <.> "o")
```

Comment:
Well, my intuition says that since everything worked fine with `stage context == Stage1` condition but started to produce incorrect results with Stage1Only flag, this has something to do with the stage when we build these objects.